README.md
# naiveproxy-docker

基于docker的naiveproxy。

<!-- TOC depthFrom:2 -->

- [1. 说明](#1-说明)
- [2. 预备工作](#2-预备工作)
- [3. 部署服务端](#3-部署服务端)
- [4. 客户端](#4-客户端)
- [5. 自定义配置](#5-自定义配置)
- [6. 版本变更](#6-版本变更)
- [7. 常见问题](#7-常见问题)
    - [7.1. 端口可以自定义吗](#71-端口可以自定义吗)

<!-- /TOC -->

## 1. 说明

镜像使用官方代码生成，利用`GitHub Actions`构建并上传到`DockerHub`。

Dockerfile：[Dockerfile](Dockerfile)

DockerHub: [DockerHub](https://hub.docker.com/repository/docker/zai7lou/naiveproxy-docker/general)

<details>
<summary>展开查看技术细节，不关心可以跳过</summary>

- 关于镜像是怎么打的

镜像先是基于go的官方镜像，安装xcaddy，然后使用xcaddy编译naiveproxy插件版的caddy。然后将caddy拷贝到debian镜像中，最后发布这个debian镜像。

这样打出来的镜像只有65M，如果不使用docker而是直接在机器上装（go + xcaddy），要1G+。

- 关于naiveproxy到底是什么

naiveproxy有客户端和服务端，这里讲的是我们部署的服务端。

naiveproxy服务端其实就是naiveproxy插件版caddy。

naiveproxy插件版caddy指的是[https://github.com/klzgrad/forwardproxy](https://github.com/klzgrad/forwardproxy)。作者通过fork原版caddy，自己实现了`forward_proxy`功能，这个就是naiveproxy代理了。

- 关于伪装

`forward_proxy`里有个`probe_resistance`指令，我们请求会先进`forward_proxy`，如果用户名密码正确，则会正常实现naiveproxy代理功能；但如果认证失败，`probe_resistance`表明不会有异常产生，而是将当前请求继续往下仍，也就是扔到我们的伪装站点（可以是反代的站点也可以是本地的文件服务）。

所以就实现了我们客户端（能提供正确的用户名和密码）去访问就是naiveproxy代理，但其他人用户浏览器访问（或认证不通过），看到的就是一个正常站点。

</details>

## 2. 预备工作

- 一个域名
- 域名已DNS到当前服务器ip
- 服务器已安装好docker环境

P.S.不需要自己生成https证书，caddy会自动生成。

## 3. 部署服务端

一键安装脚本：

```
# create a dir
mkdir -p ./naive && cd ./naive

# install
bash <(curl -sSL https://raw.githubusercontent.com/RayWangQvQ/naiveproxy-docker/main/install.sh)
```

当不指定参数时，该脚本是互动式的，运行后会提示输入相关配置信息，输入后回车即可。

![install-interaction](docs/imgs/install-interaction.png)

![install-interaction-re](docs/imgs/insatll-interaction-re.png)

当然，你也可以像下面那样，直接将参数拼接好后立即执行：

```
# create a dir
mkdir -p ./naive && cd ./naive 

# install
curl -sSL -f -o ./install.sh https://raw.githubusercontent.com/RayWangQvQ/naiveproxy-docker/main/install.sh && chmod +x ./install.sh && ./install.sh -t demo.test.tk -m zhangsan@qq.com -u zhangsan -p 1qaz@wsx --verbose
```

![install-silence](docs/imgs/install-silence.png)

参数说明：

- `-t`：host，你的域名，如`demo.test.tk`
- `-o`: cert-mode，证书模式，1为Caddy自动颁发，2为自己指定现有证书
- `-c`: cert-file，证书文件绝对路径，如`/certs/test2.zai7lou.ml.crt`
- `-k`, cert-key-file，证书key文件绝对路径，如`/certs/test2.zai7lou.ml.key`
- `-m`：mail，你的邮箱，用于自动颁发证书，如`zhangsan@qq.com`
- `-w`: http-port，http端口，默认80
- `-s`: https-port，https端口，默认443
- `-u`：user，proxy的用户名
- `-p`：pwd，proxy的密码
- `-f`：fakeHost，伪装域名，默认`https://demo.cloudreve.org`
- `--verbose`，输出详细日志
- `-h`：help，查看参数信息

容器run成功后，可以通过以下语句查看容器运行日志：

```
docker logs -f naiveproxy
```

`Ctrl + C` 可以退出日志追踪。


如果是第一次运行且选择自动颁发证书模式，颁发证书时日志可能会先ERROR飘红，别慌，等一会。

如果最后日志出现`certificate obtained successfully`字样，就是颁发成功了，可以去部署客户端了。

![success](docs/imgs/cert-suc.png)

如果颁发证书一直不成功，请检查80端口是否被占用。

部署成功后，浏览器访问域名，会展示伪装站点：

![web](docs/imgs/web.png)

## 4. 客户端

很多教程，就不说了。

|  平台   | 客户端  |
| :----:  | :----: |
|  Win    | V2RayN/Nekoray |
| Linux   | Nekoray |
| MacOS   | Nekoray |
| Android | SagerNet |
| iOS     | Shadowrocket |

## 5. 自定义配置

Caddy的配置文件`Caddyfile`已被挂载到宿主机的[./data/Caddyfile](data/Caddyfile)，想要自定义配置，比如：

- 添加proxy多用户
- 修改proxy的用户名和密码
- 更改端口
- 修改伪装站点的host

等等，都可以直接在宿主机修改该文件：

```
vim ./data/Caddyfile
```

修改完成并保存成功后，让Caddy热加载配置就可以了：

```
docker exec -it naiveproxy /app/caddy reload --config /data/Caddyfile
```

举个栗子，多用户可以直接添加`forward_proxy`，像这样：

```
{
	debug
	http_port 80
	https_port 443
	order forward_proxy before file_server
}
:443, demo.test.tk {
	tls zhangsan@qq.com
	route {
		# proxy
		forward_proxy {
			basic_auth zhangsan 1qaz@wsx
			hide_ip
			hide_via
			probe_resistance
		}
		forward_proxy {
			basic_auth lisi 1234
			hide_ip
			hide_via
			probe_resistance
		}

		# 伪装网址
		reverse_proxy you.want.com {
			header_up Host {upstream_hostport}
		}
	}
}
```

详细的配置语法可以参考Caddy的官方文档：[Caddy Doc](https://caddyserver.com/docs/)

P.S.我发现naiveproxy插件版地caddy，Caddyfile里不支持`demo.test.tk:443`的格式，必须像上面那样端口在域名前面，否则会报错。应该是适配有问题，需要注意下。

## 6. 版本变更

[CHANGELOG](CHANGELOG.md)

## 7. 常见问题
### 7.1. 端口可以自定义吗

如果使用现有证书，可以自定义；如果要Caddy自动颁发，必须占有80端口。

Caddy默认会占用80和443端口，用来管理证书，当前官方镜像并不支持更改默认端口，也就是一定需要占用80端口。

安装脚本可以选择证书模式，选择2使用现有证书，就可以不占用80和443端口了。

CHANGELOG.md
## 0.0.1
- 初始化项目
- 完成Dockerfile设计，并打出镜像
- 安装脚本实现官网版naiveproxy创建

## 0.1.0
- 完成使用现有证书功能
- 实现更改端口功能（若不使用caddy颁发证书，可不开发80端口）
- 无聊地为脚本添加一个logo
- 重构优化脚本
- 更新README

## 0.1.1
- [Fix（ #1 ）]使用现有证书时设置`auto_https`为`disable_certs`

## 0.2.1
- 发布镜像到ghcr

install.sh
#!/usr/bin/env bash
set -e
set -u
set -o pipefail

# ------------share--------------
invocation='echo "" && say_verbose "Calling: ${yellow:-}${FUNCNAME[0]} ${green:-}$*${normal:-}"'
exec 3>&1
if [ -t 1 ] && command -v tput >/dev/null; then
    ncolors=$(tput colors || echo 0)
    if [ -n "$ncolors" ] && [ $ncolors -ge 8 ]; then
        bold="$(tput bold || echo)"
        normal="$(tput sgr0 || echo)"
        black="$(tput setaf 0 || echo)"
        red="$(tput setaf 1 || echo)"
        green="$(tput setaf 2 || echo)"
        yellow="$(tput setaf 3 || echo)"
        blue="$(tput setaf 4 || echo)"
        magenta="$(tput setaf 5 || echo)"
        cyan="$(tput setaf 6 || echo)"
        white="$(tput setaf 7 || echo)"
    fi
fi

say_warning() {
    printf "%b\n" "${yellow:-}ray_naive_install: Warning: $1${normal:-}" >&3
}

say_err() {
    printf "%b\n" "${red:-}ray_naive_install: Error: $1${normal:-}" >&2
}

say() {
    # using stream 3 (defined in the beginning) to not interfere with stdout of functions
    # which may be used as return value
    printf "%b\n" "${cyan:-}ray_naive_install:${normal:-} $1" >&3
}

say_verbose() {
    if [ "$verbose" = true ]; then
        say "$1"
    fi
}

machine_has() {
    eval $invocation

    command -v "$1" >/dev/null 2>&1
    return $?
}

check_docker() {
    eval $invocation

    if machine_has "docker"; then
        docker --version
    else
        say_err "Missing dependency: docker was not found, please install it first."
        exit 1
    fi
}

# args:
# remote_path - $1
get_http_header_curl() {
    eval $invocation

    local remote_path="$1"

    curl_options="-I -sSL --retry 5 --retry-delay 2 --connect-timeout 15 "
    curl $curl_options "$remote_path" 2>&1 || return 1
    return 0
}

# args:
# remote_path - $1
get_http_header_wget() {
    eval $invocation

    local remote_path="$1"
    local wget_options="-q -S --spider --tries 5 "
    # Store options that aren't supported on all wget implementations separately.
    local wget_options_extra="--waitretry 2 --connect-timeout 15 "
    local wget_result=''

    wget $wget_options $wget_options_extra "$remote_path" 2>&1
    wget_result=$?

    if [[ $wget_result == 2 ]]; then
        # Parsing of the command has failed. Exclude potentially unrecognized options and retry.
        wget $wget_options "$remote_path" 2>&1
        return $?
    fi

    return $wget_result
}

# Updates global variables $http_code and $download_error_msg
downloadcurl() {
    eval $invocation

    unset http_code
    unset download_error_msg
    local remote_path="$1"
    local out_path="${2:-}"
    local remote_path_with_credential="${remote_path}"
    local curl_options="--retry 20 --retry-delay 2 --connect-timeout 15 -sSL -f --create-dirs "
    local failed=false
    if [ -z "$out_path" ]; then
        curl $curl_options "$remote_path_with_credential" 2>&1 || failed=true
    else
        curl $curl_options -o "$out_path" "$remote_path_with_credential" 2>&1 || failed=true
    fi
    if [ "$failed" = true ]; then
        local response=$(get_http_header_curl $remote_path)
        http_code=$(echo "$response" | awk '/^HTTP/{print $2}' | tail -1)
        download_error_msg="Unable to download $remote_path."
        if [[ $http_code != 2* ]]; then
            download_error_msg+=" Returned HTTP status code: $http_code."
        fi
        say_verbose "$download_error_msg"
        return 1
    fi
    return 0
}

# Updates global variables $http_code and $download_error_msg
downloadwget() {
    eval $invocation

    unset http_code
    unset download_error_msg
    local remote_path="$1"
    local out_path="${2:-}"
    local remote_path_with_credential="${remote_path}"
    local wget_options="--tries 20 "
    # Store options that aren't supported on all wget implementations separately.
    local wget_options_extra="--waitretry 2 --connect-timeout 15 "
    local wget_result=''

    if [ -z "$out_path" ]; then
        wget -q $wget_options $wget_options_extra -O - "$remote_path_with_credential" 2>&1
        wget_result=$?
    else
        wget $wget_options $wget_options_extra -O "$out_path" "$remote_path_with_credential" 2>&1
        wget_result=$?
    fi

    if [[ $wget_result == 2 ]]; then
        # Parsing of the command has failed. Exclude potentially unrecognized options and retry.
        if [ -z "$out_path" ]; then
            wget -q $wget_options -O - "$remote_path_with_credential" 2>&1
            wget_result=$?
        else
            wget $wget_options -O "$out_path" "$remote_path_with_credential" 2>&1
            wget_result=$?
        fi
    fi

    if [[ $wget_result != 0 ]]; then
        local disable_feed_credential=false
        local response=$(get_http_header_wget $remote_path $disable_feed_credential)
        http_code=$(echo "$response" | awk '/^  HTTP/{print $2}' | tail -1)
        download_error_msg="Unable to download $remote_path."
        if [[ $http_code != 2* ]]; then
            download_error_msg+=" Returned HTTP status code: $http_code."
        fi
        say_verbose "$download_error_msg"
        return 1
    fi

    return 0
}

# args:
# remote_path - $1
# [out_path] - $2 - stdout if not provided
download() {
    eval $invocation

    local remote_path="$1"
    local out_path="${2:-}"

    if [[ "$remote_path" != "http"* ]]; then
        cp "$remote_path" "$out_path"
        return $?
    fi

    local failed=false
    local attempts=0
    while [ $attempts -lt 3 ]; do
        attempts=$((attempts + 1))
        failed=false
        if machine_has "curl"; then
            downloadcurl "$remote_path" "$out_path" || failed=true
        elif machine_has "wget"; then
            downloadwget "$remote_path" "$out_path" || failed=true
        else
            say_err "Missing dependency: neither curl nor wget was found."
            exit 1
        fi

        if [ "$failed" = false ] || [ $attempts -ge 3 ] || { [ ! -z $http_code ] && [ $http_code = "404" ]; }; then
            break
        fi

        say "Download attempt #$attempts has failed: $http_code $download_error_msg"
        say "Attempt #$((attempts + 1)) will start in $((attempts * 10)) seconds."
        sleep $((attempts * 10))
    done

    if [ "$failed" = true ]; then
        say_verbose "Download failed: $remote_path"
        return 1
    fi
    return 0
}
# ---------------------------------

echo '  ____               _   _       _   _            '
echo ' |  _ \ __ _ _   _  | \ | | __ _(_)_(_)_   _____  '
echo ' | |_) / _` | | | | |  \| |/ _` | | | \ \ / / _ \ '
echo ' |  _ < (_| | |_| | | |\  | (_| | | |  \ V /  __/ '
echo ' |_| \_\__,_|\__, | |_| \_|\__,_| |_|   \_/ \___| '
echo '             |___/                                '

# ------------vars-----------、
gitRowUrl="https://raw.githubusercontent.com/RayWangQvQ/naiveproxy-docker/main"

host=""

certMode=""
certFile=""
certKeyFile=""
autoHttps=""

mail=""

httpPort=""
httpsPort=""

user=""
pwd=""

fakeHostDefault="https://demo.cloudreve.org"
fakeHost=""

verbose=false
# --------------------------

# read params from init cmd
while [ $# -ne 0 ]; do
    name="$1"
    case "$name" in
    -t | --host | -[Hh]ost)
        shift
        host="$1"
        ;;
    -o | --cert-mode | -[Cc]ert[Mm]ode)
        shift
        certMode="$1"
        ;;
    -c | --cert-file | -[Cc]ert[Ff]ile)
        shift
        certFile="$1"
        ;;
    -k | --cert-key-file | -[Cc]ert[Kk]ey[Ff]ile)
        shift
        certKeyFile="$1"
        ;;
    -m | --mail | -[Mm]ail)
        shift
        mail="$1"
        ;;
    -w | --http-port | -[Hh]ttp[Pp]ort)
        shift
        httpPort="$1"
        ;;
    -s | --http-port | -[Hh]ttp[Pp]ort)
        shift
        httpsPort="$1"
        ;;
    -u | --user | -[Uu]ser)
        shift
        user="$1"
        ;;
    -p | --pwd | -[Pp]wd)
        shift
        pwd="$1"
        ;;
    -f | --fake-host | -[Ff]ake[Hh]ost)
        shift
        fakeHost="$1"
        ;;
    --verbose | -[Vv]erbose)
        verbose=true
        ;;
    -? | --? | -h | --help | -[Hh]elp)
        script_name="$(basename "$0")"
        echo "Ray Naiveproxy in Docker"
        echo "Usage: $script_name [-t|--host <HOST>] [-m|--mail <MAIL>]"
        echo "       $script_name -h|-?|--help"
        echo ""
        echo "$script_name is a simple command line interface to install naiveproxy in docker."
        echo ""
        echo "Options:"
        echo "  -t,--host <HOST>         Your host, Defaults to \`$host\`."
        echo "      -Host"
        echo "          Possible values:"
        echo "          - xui.test.com"
        echo "  -m,--mail <MAIL>         Your mail, Defaults to \`$mail\`."
        echo "      -Mail"
        echo "          Possible values:"
        echo "          - mail@qq.com"
        echo "  -u,--user <USER>         Your proxy user name, Defaults to \`$user\`."
        echo "      -User"
        echo "          Possible values:"
        echo "          - user"
        echo "  -p,--pwd <PWD>         Your proxy password, Defaults to \`$pwd\`."
        echo "      -Pwd"
        echo "          Possible values:"
        echo "          - 1qaz@wsx"
        echo "  -f,--fake-host <FAKEHOST>         Your fake host, Defaults to \`$fakeHost\`."
        echo "      -FakeHost"
        echo "          Possible values:"
        echo "          - https://demo.cloudreve.org"
        echo "  -?,--?,-h,--help,-Help             Shows this help message"
        echo ""
        exit 0
        ;;
    *)
        say_err "Unknown argument \`$name\`"
        exit 1
        ;;
    esac
    shift
done

read_var_from_user() {
    eval $invocation

    # host
    if [ -z "$host" ]; then
        read -p "请输入域名(如demo.test.tk):" host
    else
        say "域名: $host"
    fi

    # cert
    if [ -z "$certMode" ]; then
        read -p "请输入证书模式(1.Caddy自动颁发；2.使用现有证书。默认1):" certMode
        if [ -z "$certMode" ]; then
            certMode="1"
        fi
    fi

    if [ "$certMode" == "1" ]; then
        # say "certMode: $certMode（由Caddy自动颁发）"
        say_warning "自动颁发证书需要开放80端口给Caddy使用，请确保80端口开放且未被占用"
        httpPort="80"
        # email
        if [ -z "$mail" ]; then
            read -p "请输入邮箱(如test@qq.com):" mail
        else
            say "邮箱: $mail"
        fi
    else
        # say "certMode: 2（使用现有证书）"
        autoHttps="auto_https disable_certs"
        if [ -z "$certKeyFile" ]; then
            read -p "请输入证书key文件路径:" certKeyFile
        else
            say "证书key: $certKeyFile"
        fi

        if [ -z "$certFile" ]; then
            read -p "请输入证书文件路径:" certFile
        else
            say "证书文件: $certFile"
        fi
    fi

    # port
    if [ -z "$httpPort" ]; then
        if [ $certMode == "2" ]; then
            say "使用现有证书模式允许使用非80的http端口"
            read -p "请输入Caddy的http端口(如8080, 默认80):" httpPort
            if [ -z "$httpPort" ]; then
                httpPort="80"
            fi
        else
            httpPort="80"
            say "Http端口: $httpPort"
        fi
    else
        say "httpPort: $httpPort"
    fi

    if [ -z "$httpsPort" ]; then
        read -p "请输入https端口(如8043, 默认443):" httpsPort
        if [ -z "$httpsPort" ]; then
            httpsPort="443"
        fi
    else
        say "Https端口: $httpsPort"
    fi

    if [ -z "$user" ]; then
        read -p "请输入节点用户名(如zhangsan):" user
    else
        say "节点用户名: $user"
    fi

    if [ -z "$pwd" ]; then
        read -p "请输入节点密码(如1qaz@wsx):" pwd
    else
        say "节点密码: $pwd"
    fi

    if [ -z "$fakeHost" ]; then
        read -p "请输入伪装站点地址(默认$fakeHostDefault):" fakeHost
        if [ -z "$fakeHost" ]; then
            fakeHost=$fakeHostDefault
        fi
    else
        say "伪装站点地址: $fakeHost"
    fi
}

# 下载docker-compose文件
download_docker_compose_file() {
    eval $invocation

    rm -rf ./docker-compose.yml
    download $gitRowUrl/docker-compose.yml docker-compose.yml
}

# 配置docker-compose文件
replace_docker_compose_configs() {
    eval $invocation

    # replace httpPort
    sed -i 's|<httpPort>|'"$httpPort"'|g' ./docker-compose.yml

    # replace httpsPort
    sed -i 's|<httpsPort>|'"$httpsPort"'|g' ./docker-compose.yml

    # certs
    if [ "$certMode" == "2" ]; then
        sed -i 's|<certVolumes>|'-" $certFile":"$certFile"'|g' ./docker-compose.yml
        sed -i 's|<certKeyVolumes>|'-" $certKeyFile":"$certKeyFile"'|g' ./docker-compose.yml
    else
        sed -i 's|<certVolumes>| |g' ./docker-compose.yml
        sed -i 's|<certKeyVolumes>| |g' ./docker-compose.yml
    fi

    say "Docker compose file:"
    cat ./docker-compose.yml
}

# 下载data
download_data_files() {
    eval $invocation

    mkdir -p ./data

    # entry
    rm -rf ./data/entry.sh
    download $gitRowUrl/data/entry.sh ./data/entry.sh

    # Caddyfile
    rm -rf ./data/Caddyfile
    download $gitRowUrl/data/Caddyfile ./data/Caddyfile
}

# 配置Caddyfile
replace_caddyfile_configs() {
    eval $invocation

    # debug
    debug=""
    if [ $verbose = true ]; then
        debug="debug"
    fi
    sed -i 's|<debug>|'"$debug"'|g' ./data/Caddyfile

    # replace host
    sed -i 's|<host>|'"$host"'|g' ./data/Caddyfile

    # replace mail
    sed -i 's|<mail>|'"$mail"'|g' ./data/Caddyfile

    # cert_file
    sed -i 's|<cert_file>|'"$certFile"'|g' ./data/Caddyfile

    # cert_key_file
    sed -i 's|<cert_key_file>|'"$certKeyFile"'|g' ./data/Caddyfile

    # replace httpPort
    sed -i 's|<httpPort>|'"$httpPort"'|g' ./data/Caddyfile

    # replace httpsPort
    sed -i 's|<httpsPort>|'"$httpsPort"'|g' ./data/Caddyfile

    # auto_https
    sed -i 's|<autoHttps>|'"$autoHttps"'|g' ./data/Caddyfile

    # replace user
    sed -i 's|<user>|'"$user"'|g' ./data/Caddyfile

    # replace pwd
    sed -i 's|<pwd>|'"$pwd"'|g' ./data/Caddyfile

    # replace fakeHost
    sed -i 's|<fakeHost>|'"$fakeHost"'|g' ./data/Caddyfile

    say "Caddyfile:"
    cat ./data/Caddyfile
}

# 运行容器
runContainer() {
    eval $invocation

    say "Try to run docker container:"
    {
        docker compose version && docker compose up -d
    } || {
        docker-compose version && docker-compose up -d
    } || {
        certsV=""
        if [ "$certMode" == "2" ]; then
            certsV="-v $certFile:certFile -v $certKeyFile:$certKeyFile"
        fi
        docker run -itd --name naiveproxy \
        --restart=unless-stopped \
        -p $httpPort:$httpPort \
        -p $httpsPort:$httpsPort \
        -v $PWD/data:/data \
        -v $PWD/share:/root/.local/share $certsV \
        zai7lou/naiveproxy-docker bash /data/entry.sh
    }
}

# 检查容器运行状态
check_result() {
    eval $invocation

    docker ps --filter "name=naiveproxy"

    containerId=$(docker ps -q --filter "name=^naiveproxy$")
    if [ -n "$containerId" ]; then
        echo ""
        echo "==============================================="
        echo "Congratulations! 恭喜！"
        echo "创建并运行naiveproxy容器成功。"
        echo ""
        echo "请使用浏览器访问'https://$host:$httpsPort'，验证是否可正常访问伪装站点"
        echo "如果异常，请运行'docker logs -f naiveproxy'来追踪容器运行日志, 随后可以点击 Ctrl+c 退出日志追踪"
        echo ""
        echo "然后你可以使用客户端连接你的节点了："
        echo "naive+https://$user:$pwd@$host:$httpsPort#naive"
        echo "Enjoy it~"
        echo "==============================================="
    else
        echo ""
        echo "请查看运行日志，确认容器是否正常运行，点击 Ctrl+c 退出日志追踪"
        echo ""
        docker logs -f naiveproxy
    fi
}

main() {
    check_docker
    read_var_from_user

    download_docker_compose_file
    replace_docker_compose_configs

    download_data_files
    replace_caddyfile_configs

    runContainer

    check_result
}

main
